text/0000-safety-tags.md

```diff
@@ -0,0 +1,485 @@
+pub unsafe fn read<T>(ptr: *const T) -> T { ... }
+
+#[clippy::safety { ValidPtr, Aligned, Initialized }] // callsite
+unsafe { read(ptr) };
+```
+
+We can also attach comments for a tag or a group of tags to clarify how safety requirements are met:
+
+```rust
+#[clippy::safety {
+  ValidPtr, Aligned, Initialized: "addr range p..p+n is properly initialized from aligned memory"
+  InBounded, ValidNum: "`n` won't exceed isize::MAX here, so `p.add(n)` is fine";
+}]
+for _ in 0..n {
+    unsafe {
+        c ^= p.read();
+        p = p.add(1);
+    }
+}
+```
+
+[tool attribute]: https://doc.rust-lang.org/reference/attributes.html#tool-attributes
+[`ptr::read`]: https://doc.rust-lang.org/std/ptr/fn.read.html
+
+Every safety tag declared on a function must appear in `#[clippy::safety { ... }]` together with an
+optional reason; any omission triggers a warning-by-default diagnostic that lists the missing tags
+and explains each one:
+
+```rust
+unsafe { ptr::read(ptr) }
+```
+
+```rust
+warning: `ValidPtr`, `Aligned`, `Initialized` tags are missing. Add them to `#[clippy::safety { }]`.
+   --> file.rs:xxx:xxx
+    |
+LLL | unsafe { ptr::read(ptr) }
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^ This unsafe call requires these safety tags.
+    |
+    = NOTE: See core::ptr::invariants::ValidPtr
+    = NOTE: See core::ptr::invariants::Aligned
+    = NOTE: See core::ptr::invariants::Initialized
+```
+
+The process of verifying whether a tag is present is referred to as tag discharge.
+
+Note that it's allowed to discharge tags of unsafe callees onto the unsafe caller for unsafe
+delegation or propogation:
+
+```rust
+#[clippy::safety { ValidPtr, Aligned, Initialized }] // ✅
+unsafe fn constructor<T>() -> T {
+    unsafe { read(...) }
+}
+```
+
+## Safety Tags as Ordinary Items
+
+Before tagging a function, we must declare them as ordinary items with `#[clippy::safety::tag]` such
+as [uninhabited] types or plain functions:
+
+```rust
+#[clippy::safety::tag]
+enum ValidPtr {}
+
+#[clippy::safety::tag]
+fn Aligned() {}
+```
+
+Tags live in their own [type namespace] carry item-level [scopes] and obey [visibility] rules,
+keeping the system modular and collision-free. Since they are never referenced directly as real
+items, we propose importing them uses a dedicated syntax: inner-styled or outer-styled 
+`clippy::safety::r#use` tool attribute on modules:
+
+```rust
+#![clippy::safety::r#use { UseTree })] // {} signifies a delimiter here, thus () also works
+```
+
+[`UseTree`] follows the exact grammar of the `use` declaration. Some examples:
+
+```rust
+#[clippy::safety::r#use { core::ptr::invariants::* }]
+mod foo;
+
+mod bar {
+    #![clippy::safety::r#use { core::ptr::invariants::{ValidPtr, Aligned} }]
+}
+```
+
+That's to say:
+* Tags declared or re-exported in the current module are automatically in scope: no import required.
+* Tags from other modules must be brought in with the inner-tool attribute shown above.
```

Comment:
we can't `#[clippy::safety { core::ptr::invariants::Aligned }] unsafe ...` directly without using `r#use`?

Reply:
We can restore `safety` without `r#use` to import tags if `safety::requires` and `safety::checked` are syntax for tagging.
[Previous discussion to why `r#use`.](https://github.com/rust-lang/rfcs/pull/3842#discussion_r2246032076)

Reply:
Nit: `safety::r#use` won't happen on unsafe fns or expressions. So 

```diff
-#[clippy::safety { core::ptr::invariants::Aligned }]
+#[clippy::safety { core::ptr::invariants::Aligned }] unsafe ... 
```

Reply:
But to be clearer, how about `#![clippy::safety::import(...)]` to avoid `use` keyword? For now, we have these syntax:

```rust
#[clippy::safety::import(...)]       # import tag items
#[clippy::safety::declare_tag(...)]  # declare a tag on items
#[clippy::safety::requires(...)]     # define tags on unsafe functions
#[clippy::safety::checked(...)]      # discharge tags on expressions containing unsafe calls
```

Reply:
@zjp-CN my primary concern is that this item is saying

> <ins>① Tags from other modules</ins> must be brought in with the inner-tool <ins>② attribute shown above</ins>.

so it sounds like if you don't use that "② attribute" which is `#[r#use]`/`#[import]` you can't use any "① tags from other modules" even via full path.